pkg/deploy/gateway.go

```diff
@@ -0,0 +1,498 @@
+		return err
+	}
+
+	serverConfig := getGatewayServerConfigSpec(instance)
+	if err := sync(instance, clusterAPI, &serverConfig, configMapDiffOpts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func deleteAll(instance *orgv1.CheCluster, clusterAPI ClusterAPI) error {
+	deployment := appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GatewayServiceName,
+			Namespace: instance.Namespace,
+		},
+	}
+	if err := delete(clusterAPI, &deployment); err != nil {
+		return err
+	}
+
+	serverConfig := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      gatewayServerConfigName,
+			Namespace: instance.Namespace,
+		},
+	}
+	if err := delete(clusterAPI, &serverConfig); err != nil {
+		return err
+	}
+
+	traefikConfig := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "che-gateway-config",
+			Namespace: instance.Namespace,
+		},
+	}
+	if err := delete(clusterAPI, &traefikConfig); err == nil {
+		return err
+	}
+
+	roleBinding := rbac.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GatewayServiceName,
+			Namespace: instance.Namespace,
+		},
+	}
+	if err := delete(clusterAPI, &roleBinding); err == nil {
+		return err
+	}
+
+	role := rbac.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GatewayServiceName,
+			Namespace: instance.Namespace,
+		},
+	}
+	if err := delete(clusterAPI, &role); err == nil {
+		return err
+	}
+
+	sa := corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GatewayServiceName,
+			Namespace: instance.Namespace,
+		},
+	}
+	if err := delete(clusterAPI, &sa); err == nil {
+		return err
+	}
+
+	return nil
+}
+
+// sync syncs the blueprint to the cluster in a generic (as much as Go allows) manner.
+func sync(instance *orgv1.CheCluster, clusterAPI ClusterAPI, blueprint metav1.Object, diffOpts cmp.Option) error {
+	blueprintObject, ok := blueprint.(runtime.Object)
+	if !ok {
+		return fmt.Errorf("object %T is not a runtime.Object. Cannot sync it", blueprint)
+	}
+
+	key := client.ObjectKey{Name: blueprint.GetName(), Namespace: blueprint.GetNamespace()}
+
+	actual := blueprintObject.DeepCopyObject()
+
+	if getErr := clusterAPI.Client.Get(context.TODO(), key, actual); getErr != nil {
+		if statusErr, ok := getErr.(*errors.StatusError); !ok || statusErr.Status().Reason != metav1.StatusReasonNotFound {
+			return getErr
+		}
+		actual = nil
+	}
+
+	kind := blueprintObject.GetObjectKind().GroupVersionKind().Kind
+
+	if actual == nil {
+		logrus.Infof("Creating a new object: %s, name %s", kind, blueprint.GetName())
+		err := clusterAPI.Client.Create(context.TODO(), blueprintObject)
+		if err != nil {
+			return err
+		}
+	} else {
+		actualMeta := actual.(metav1.Object)
+
+		diff := cmp.Diff(actual, blueprint, diffOpts)
+		if len(diff) > 0 {
+			logrus.Infof("Updating existing object: %s, name: %s", kind, actualMeta.GetName())
+			fmt.Printf("Difference:\n%s", diff)
+
+			err := clusterAPI.Client.Delete(context.TODO(), actual)
```

Comment:
I use Delete/Create approach to update the object for Ingress/Routes/Services only. I don't remember the exact reason. 
For another objects I use `Update`. 
I don't think that Delete/Update is good for deployment. Deleting requires stopping pods, so we will have AlreadyExists error on Create below. Besides that I check if pods are running
https://github.com/eclipse/che-operator/blob/master/pkg/deploy/deployment.go#L96-L103
In another word I am afraid if deployment is changed we have a lot errors in operator logs.